docs/content/sips/xxx-spin-plugins.md

```diff
@@ -0,0 +1,154 @@
+title = "SIP xxx - Plugin System for Spin"
+template = "main"
+date = "2022-08-09T13:22:30Z"
+---
+
+Summary: Plugin system for spin.
+
+Owner: karthik.ganeshram@fermyon.com
+
+Created: August 9, 2022 
+
+## Background 
+
+As the functionality fo spin gets extended, there will be a point where every feature will not be required/used by every user. An example of this would be different triggers for spin along with possible subcommands to add more functionality to spin. Therefore it would make sense to be able to add features as plugins allowing for addition of only the required features witout modifying the spin binary.
+
+## Proposal
+
+Create a new subcommand for spin called `spin plugin` which will have two further subcommands namely
+
+- `install`
+- `uninstall`
+
+### Types of plugins
+
+![Classification of types of plugins](https://i.imgur.com/Fo7EGPQ.png)
+
+- Standalone Binary Plugins (external subcommands) - These work to add features to spin and do not need any use of the spin binary.
+- Internal Spin Plugins - These are plugins like triggers that are used by the spin application. 
+- Spin dependant User Plugins - These user invoked plugins require some functionality provided by spin.
+
+### Proposed workflow
+
+A plugin can be installed using the following command.
+
+```bash
+$ spin plugin install <name_of_plugin>
+The following $plugin from <source_of_plugin> will be installed. 
+Make sure you trust this source! Continue? <y/N> 
+```
+
+This installs the plugin to a to a spin managed directory based on the operating system.
+
+In the case of standaone binary plugins and spin dependant user plugins, the plugin can be invoked as 
+
+```bash
+$ spin $plugin <args_to_plugin>
+```
+
+In the case of spin Dependant Internal Plugins, the user does not directly invoke the command, these plugins are directly invoked by spin where required (i.e) like triggers based on the application manifest. [(Spin Trigger Executor)](https://spin.fermyon.dev/sips/003-trigger-executors.md)
+
+To uninstall plugins, the following command can be used.
+
+```
+spin uninstall $plugin
+```
+
+This will uninstall the executable from the spin managed directory.
+
+
+### Execution of plugins
+
+The execution of the plugins works differently based on the type of plugin.
+
+- **Standalone Binary Plugin** 
+  - Once the plugin is invoked using `spin $plugin <args_to_plugin>`, the external subcommand is invoked as `spin-$plugin <args>`.
+- **Spin Dependant User Plugins** 
+  - Spin first provides all the functionality available in it required by the plugin and then proceeds to call the plugin along with the required parameters.
```

Comment:
I will try to make it clearer with some made-up example plugins. 

**Standalone Binary Plugin** - These are just basic binaries that can add functionality to spin (eg) `spin routes` just displays the HTTP routes for a given application from `spin.toml`.

**Spin-based User Plugins** - A made-up example of this would be `spin deploy`  where the user invokes the command but as a plugin, it would require some functionality from spin. Therefore `deploy` is a type of command and it could have multiple plugins each of which deploys to different services. 

**Spin-based Internal Plugins** - An example of this would be triggered, where the user does not directly call for the triggers but uses them by specifying them in the application manifest. Loaders would be another example of this.


And yes, Spin will be able to distinguish between the types of plugins from the name of the plugin.

Both Categories of Spin-based plugins will have keywords signifying types and all the types will be known ahead of time as each type of plugin follows a different execution path. The categories will be split into `internal` and `user` plugins.

An example of this would be a time trigger plugin called `spin-trigger-time` where the type of plugin is `trigger` which will be contained in a list of internal plugins.

Reply:
These are great examples - thanks!

So for both categories of "dependent plugin", the behaviour is:

* Spin looks at the name of the plugin binary (e.g. `spin-foo-bar` and extracts the _second_ component.  This identifies a _type_ of plugin and must be a well known string.  [Or is the type derived in some other way, e.g. from the Spin command being issued, and the name of the plugin inferred from that.]
* Based on that type, Spin knows the contract for calling that type of plugin, e.g. what files must be placed where, what environment variables must be set, what command line flags must be provided, etc.
* Spin performs that type-based preparation.
* Spin invokes the plugin.

Is that right?  If so, it would be super useful (to me at least) to capture that flow.

User and internal DPs, then, have the exact same flow at the technical level, but differ in how Spin decides to launch them.  E.g. a hypothetical AWS deployment plugin would be chosen explicitly by the command line user issuing `spin deploy aws` - Spin knows how to call `deploy` plugins but just blindly calls them by name.  But trigger plugins are called implicitly based on the content of `spin.toml`, and that's what makes them internal.  Is that right?

Reply:
Yes, This is exactly the flow.

The type of plugin is inferred from the second component of the plugin name and it is a well-known string.